pkg/cmd/healthprobe/healthprobe.go

```diff
@@ -0,0 +1,316 @@
+var shutdownSignals = []os.Signal{os.Interrupt, syscall.SIGTERM}
+
+func (h *healthProbeOpts) Validate(_ context.Context) error {
+	if len(h.clientKeyFile) == 0 {
+		return errors.New("missing required flag: --key-file")
+	}
+	if len(h.clientCertFile) == 0 {
+		return errors.New("missing required flag: --cert-file")
+	}
+	if len(h.clientCACertFile) == 0 {
+		return errors.New("missing required flag: --cacert-file")
+	}
+	return nil
+}
+
+func (h *healthProbeOpts) Run(ctx context.Context) error {
+	defer utilruntime.HandleCrash()
+	ctx, cancel := context.WithCancel(ctx)
+
+	// handle teardown
+	shutdownHandler := make(chan os.Signal, 2)
+	signal.Notify(shutdownHandler, shutdownSignals...)
+	go func() {
+		select {
+		case <-shutdownHandler:
+			klog.Infof("Received SIGTERM or SIGINT signal, shutting down.")
+			close(shutdownHandler)
+			cancel()
+		case <-ctx.Done():
+			klog.Infof("Context has been cancelled, shutting down.")
+			close(shutdownHandler)
+			cancel()
+		}
+	}()
+
+	// enable profiler by default
+	defer serviceability.Profile("web").Stop()
+	serviceability.StartProfiler()
+
+	healthCheck, err := h.newHealthCheck()
+	if err != nil {
+		return err
+	}
+
+	healthCheckProbe := healthCheck.Probe()
+	if err := wait.PollImmediateUntil(h.interval, func() (bool, error) { return healthCheckProbe() }, ctx.Done()); err != nil {
+		// probe is shutting down report last status of existing disruption
+		if err == wait.ErrWaitTimeout && healthCheck.Disruption {
+			healthCheck.lg.Warn("health probe terminated",
+				zap.String("pod", healthCheck.podName),
+				zap.Bool("available", false),
+				zap.Duration("duration", time.Since(healthCheck.DisruptionStart)), // duration
+			)
+		}
+	}
+	healthCheck.lg.Info("health probe is shutting down",
+		zap.String("pod", healthCheck.podName),
+	)
+	return nil
+}
+
+func (h *healthProbeOpts) newHealthCheck() (*healthCheck, error) {
+	if h.enableLogRotation {
+		if err := setupLogRotation(h.logOutputs, h.LogRotationConfigJSON); err != nil {
+			return nil, err
+		}
+	}
+	lg, err := getZapLogger(loglevelToZap(h.logLevel), h.logOutputs, h.enableLogRotation)
+	if err != nil {
+		return nil, err
+	}
+	tlsInfo := transport.TLSInfo{
+		CertFile:      h.clientCertFile,
+		KeyFile:       h.clientKeyFile,
+		TrustedCAFile: h.clientCACertFile,
+	}
+	tr, err := transport.NewTransport(tlsInfo, h.dialTimeout)
+	if err != nil {
+		return nil, err
+	}
+	client := http.Client{
+		Transport: tr,
+	}
+
+	healthCheck := &healthCheck{
+		client:            client,
+		endpoint:          h.Endpoint,
+		lg:                lg,
+		interval:          h.interval,
+		podName:           h.podName,
+		staticPodVersion:  h.staticPodVersion,
+		enableLogRotation: h.enableLogRotation,
+	}
+	return healthCheck, nil
+}
+
+func (h *healthCheck) Probe() HealthCheckFunc {
+	h.lg.Warn("starting health probe",
+		zap.String("pod", h.podName),
+	)
+	return func() (bool, error) {
+		isHealthy, _ := checkHealth(h.client, h.endpoint, h.podName, h.lg)
```

Comment:
Why structure like this instead of simply

`func (h *healthCheck) CheckQuorumRead()` and have the call site do

```go
func() (bool, error){
  h.CheckQuorumRead(ctx)
  return false, nil
}
```